src/core/maps/cell_occupancy_estimator.h

```diff
@@ -47,6 +56,73 @@ struct Occupancy {
 struct Beam {
   double x_st, y_st; ///< Coordinates of a start of the beam.
   double x_end, y_end; ///< Coordinates where the beam reaches an obstacle.
+
+  //methods
+  bool touches_rect(const Rectangle& bnds) const {
+    return equal(x_st, x_end, bnds.left) || equal(y_st, y_end, bnds.bot) ||
+           equal(x_st, x_end, bnds.right) || equal(y_st, y_end, bnds.top);
+  }
+
+  Beam operator- () const {
+    return (*this).revert();
+  }
+
+
+
+  bool intersects(const Rectangle& bnds) const {
+    bool is_inside_rect = bnds.is_inside(x_st, y_st) ||
+                          bnds.is_inside(x_end, y_end);
+    if (is_inside_rect) {
+      return true;
+    }
+
+    bool result = false;
+    Ray ray(x_st, y_st, x_end - x_st, y_end - y_st);
+    Intersections intrs = ray.find_intersections(bnds);
+    for (auto intersection : intrs) {
+      result |= bnds.is_on_border(intersection.x, intersection.y);
+    }
+    return result;
+  }
+
+  bool encounters(const Rectangle& bnds) const {
```

Comment:
this method checks whether a beam stops on a border of a cell and not passes through.
may be there is some misunderstanding: touches_rect means that the beam contains one border (or a piece of border). Encounters means that the beam stops on border, but no one point of beam contains inside a cell